File: iOS/ContentExtension/NotificationViewController.m
//
//  NotificationViewController.m
//  CustomNotificationContent
//
//  Created by qgraph on 14/09/16.
//  Copyright © 2016 QuantumGraph. All rights reserved.
//

#import "NotificationViewController.h"
#import <UserNotifications/UserNotifications.h>
#import <UserNotificationsUI/UserNotificationsUI.h>
#import "QGNotificationSdk.h"

@interface NotificationViewController () <UNNotificationContentExtension, iCarouselDelegate, iCarouselDataSource>

@property (nonatomic, weak) IBOutlet iCarousel *carousel;

@end

static NSString *APP_GROUP = @"group.com.company.product.notification";

@implementation NotificationViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    //configure your action button titles
    // for ios 12 only open app button is shown, however for ios 11 and below, open app and next button both are shown.
    // for ios 12 user can also scroll carousel/slider using interaction
    [[QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP] viewDidLoadWithCarousel:_carousel withActionNextTitle:nil withActionOpenAppTitle:nil];
}

- (void)didReceiveNotification:(UNNotification *)notification
{
    [[QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP] didReceiveNotification:notification];
}

- (void)didReceiveNotificationResponse:(UNNotificationResponse *)response completionHandler:(void (^)(UNNotificationContentExtensionResponseOption))completion
{
    QGNotificationSdk *qgsdk = [QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP];
    [qgsdk didReceiveNotificationResponse:response withContext:(NSExtensionContext *)self.extensionContext completionHandler:completion];
}

///--------------------------------
#pragma mark - iCarousel Delegate
///--------------------------------

- (NSInteger)numberOfItemsInCarousel:(iCarousel *)carousel
{
    return [[QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP] numberOfItemsInCarousel:carousel];
}

- (UIView *)carousel:(iCarousel *)carousel viewForItemAtIndex:(NSInteger)index reusingView:(UIView *)view
{
    return [[QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP] carousel:carousel viewForItemAtIndex:index reusingView:view];
}

- (CGFloat)carousel:(iCarousel *)carousel valueForOption:(iCarouselOption)option withDefault:(CGFloat)value
{
    return [[QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP] carousel:carousel valueForOption:option withDefault:value];
}

- (CGFloat)carouselItemWidth:(iCarousel *)carousel {
    return [[QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP] carouselItemWidth:carousel];
}

- (void)carousel:(iCarousel *)carousel didSelectItemAtIndex:(NSInteger)index
{
    if (@available(iOS 12.0, *)) {
        [[QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP] carousel:carousel didSelectItemAtIndex:index withExtensionContext:self.extensionContext];
    } else {
        // Fallback on earlier versions
    }
}

- (void)carouselDidEndDragging:(iCarousel *)carousel willDecelerate:(BOOL)decelerate {
    if (@available(iOS 12.0, *)) {
        [[QGNotificationSdk sharedInstanceWithAppGroup:APP_GROUP] carouselDidEndDragging:carousel willDecelerate:decelerate];
    }
}

- (void)dealloc
{
    self.carousel.delegate = nil;
    self.carousel.dataSource = nil;
}

@end

File: iOS/Swift-5.2.3/QGNotificationContentExtension/NotificationViewController.swift
//
//  NotificationViewController.swift
//  ContentExtension
//
//  Created by Appier on 2019/8/23.
//  Copyright © 2019 Appier. All rights reserved.
//

import UIKit
import UserNotifications
import UserNotificationsUI

class NotificationViewController: UIViewController, UNNotificationContentExtension, iCarouselDelegate, iCarouselDataSource {

    let APP_GROUP:String = "group.com.appier.NickTest.notification"
    
    @IBOutlet weak var carousel: iCarousel!
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        _ = QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP).viewDidLoad(with: carousel)
    }
    
    func didReceive(_ notification: UNNotification) {
        QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP).didReceive(notification)
    }
    
    func didReceive(_ response: UNNotificationResponse, completionHandler completion: @escaping (UNNotificationContentExtensionResponseOption) -> Void) {
        if let theContext = self.extensionContext {
            let qgsdk = QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP)
            qgsdk.didReceive(response, with: theContext, completionHandler: completion)
        }
    }
    
    func numberOfItems(in carousel: iCarousel) -> Int {
        return QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP).numberOfItems(in: carousel)
    }
    
    func carousel(_ carousel: iCarousel, viewForItemAt index: Int, reusing view: UIView?) -> UIView {
        return QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP).carousel(carousel, viewForItemAt: index, reusing: view)
    }
    
    func carousel(_ carousel: iCarousel, valueFor option: iCarouselOption, withDefault value: CGFloat) -> CGFloat {
        return QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP).carousel(carousel, valueFor: option, withDefault: value)
    }
    
    func carouselItemWidth(_ carousel: iCarousel) -> CGFloat {
        return QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP).carouselItemWidth(carousel)
    }
    
    func carousel(_ carousel: iCarousel, didSelectItemAt index: Int) {
        if #available(iOS 12.0, *) {
            if let theContext = self.extensionContext {
                QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP).carousel(carousel, didSelectItemAt: index, with: theContext)
            }
        } else {
            //Fallback on earlier versions
        }
    }
    
    func carouselDidEndDragging(_ carousel: iCarousel, willDecelerate decelerate: Bool) {
        if #available(iOS 12.0, *) {
            QGNotificationSdk.sharedInstance(withAppGroup: APP_GROUP).carouselDidEndDragging(carousel, willDecelerate: decelerate)
        }
    }
    
    deinit {
        self.carousel.delegate = nil
        self.carousel.dataSource = nil
    }
    
}
